[WIP] [AI-11] (P)PCA implementation

DanielRrr opened the pull request.

Haskell-mouse committed:
AI-11 implement PCA and update dependencies

DanielRrr commented on src/GPLVM/PCA.hs:
```diff
@@ -0,0 +1,39 @@
+    , covariance  :: Covariance Double
+    , eigenVectors :: EigenVectors Double
+    , eigenValues :: Vector F Double
+    , finalData :: Matrix D Dbouble
```
`Dbouble`? what's that? Some new numerical data type that widely used for PCA? :peka:

DanielRrr commented on src/GPLVM/PCA.hs:
```diff
@@ -0,0 +1,39 @@
+      eigenVectors =
+        if desiredDimentions >= y
+        then Matrix eigen
+        else Matrix $ fromForeignPtr (Z :. desiredDimentions :. x) $ toForeignPtr eigenVec
```
`desiredDimentions` -> `desiredDimensions`

DanielRrr commented on src/GPLVM/Types.hs:
```diff
@@ -1,46 +1,59 @@
 import qualified Numeric.Matrix as M
 
-type Distribution a = Matrix a -> a
+type Distribution r a = Matrix D r a -> a
```
`type Distribution r a = Matrix D r a -> a` -> `type Distribution r a = Matrix r a -> a`

DanielRrr commented on src/GPLVM/Types.hs:
```diff
@@ -1,46 +1,59 @@
+type Distribution r a = Matrix D r a -> a
 
-type Mean a = Matrix a -> a
+type Mean r a = Matrix D r a -> a
```
`type Mean r a = Matrix D r a -> a` -> `type Mean r a = Matrix r a -> a`

DanielRrr commented on src/GPLVM/Types.hs:
```diff
@@ -1,46 +1,59 @@
 
+type Covariance a = Herm a
 
+newtype Matrix D r a = Matrix D (Array r DIM2 a)
```
`newtype Matrix D r a = Matrix D (Array r DIM2 a)` -> `newtype Matrix r a = Matrix D (Array r DIM2 a)`

DanielRrr commented on src/GPLVM/Types.hs:
```diff
@@ -1,46 +1,59 @@
 
-newtype LatentSpacePoints a = LatentSpacePoints { _unLatentSpacePoints :: Matrix a }
-    deriving Eq
+newtype LatentSpacePoints a = LatentSpacePoints { _unLatentSpacePoints :: Matrix D D a }
```
Leave `Matrix D a`

DanielRrr commented on src/Math/Matrix.hs:
```diff
@@ -1,3 +1,6 @@
 
 import Universum
+
+test :: Int 
```
remove this `test`

DanielRrr reviewed (commented):
Enjoy your diss ❤️

Commit:
[AI-11] fixes in PCA implementation

Commit:
[AI-11] Improve and test PCA.

DanielRrr commented on src/GPLVM/PCA.hs:
```diff
@@ -0,0 +1,44 @@
+    , _eigenVectors :: EigenVectors Double
+    , _eigenValues  :: Vector F Double
+    , _finalData    :: Matrix D Double
+    , _restoredData :: Matrix D Double
```
can we remove this field with restored data?

Haskell-mouse replied:
Why? It is one of the main results of the algorithm.

DanielRrr commented on stack.yaml:
```diff
@@ -1,11 +1,11 @@
 
 - git: https://github.com/adscib/monad-bayes
   commit: 647ba7cb5a98ae028600f3d828828616891b40fb
+- /root/repa-linear-algebra
```
I would like to add `repa-linear-algebra` dependency via git link to your fork

DanielRrr replied:
BTW, add please our test file that we've created yesterday.

Haskell-mouse replied:
ok. I'll push the changes and replace that )

DanielRrr reviewed (commented):
Looks good for me. Consider please my small comments

Commit:
[AI-11] Fix stack.yaml

Commit:
[AI-11] basic GPVLM impl (wip)

Commit:
[AI-14] updates in src/GPLVM/GaussianProcessLV.hs

Commit:
[AI-14] debug gp

Commit:
[AI-11] Add PPCA procedure

1. Added PPCA data type and corresponding functions.
2. Removed Matrix newtype (i tried to use it for checking dimensions in complie-time). Now
   Matrix is just a type synonym.

Commit:
Done tests of step one

Commit:
[AI-11] PPCA now works with full data.

Also implemented PPCA with missed data points, but it is needed to
debug it. The same problem as it was with usual PPCA.

Also added many functions to Utils. Think that these functions
should be added to REPA due to its usefullness.

Commit:
[AI-20] AI-20 Write matrix with type-safe dimensions and implement PPCA using it

1. create proper data types and utility functions for type-safe dimensions
2. debug PPCA with missed data (WIP)
3. Write PPCA using new data type

Commit:
[AI-20] Add forgotten file
